ocaml/alerts/expiry_alert.ml

```diff
@@ -0,0 +1,126 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module XenAPI = Client.Client
+module Date = Xapi_stdext_date.Date
+
+let seconds_per_day = 3600. *. 24.
+
+let days_until_expiry epoch expiry =
+  (expiry /. seconds_per_day) -. (epoch /. seconds_per_day)
+
+let all_messages rpc session_id =
+  XenAPI.Message.get_all_records ~rpc ~session_id
+
+let related_messages msg_obj_uuid related_message_name_list all_msgs =
+  List.filter
+    (fun (_ref, record) ->
+      let related name = List.mem name related_message_name_list in
+      record.API.message_obj_uuid = msg_obj_uuid
+      && related record.API.message_name
+    )
+    all_msgs
+
+let message_body msg expiry =
+  Printf.sprintf "<body><message>%s</message><date>%s</date></body>" msg
+    (Date.to_string expiry)
+
+let expired_message obj = Printf.sprintf "The %s has expired." obj
+
+let expiring_message obj = Printf.sprintf "The %s is expiring soon." obj
+
+let generate_alert now expiry expired_message_id expiring_conditions
+    alert_obj_description =
+  let remaining_days =
+    days_until_expiry (Date.to_float now) (Date.to_float expiry)
+  in
+  let sorted_alert_conditions =
```

Comment:
Something like this would be able to eliminate the `List.sort` and `List.find_opt`:
```
let rec get_most_critical = function
  | [], curr ->
      curr
  | ((days, _) as cond) :: t, None ->
      if days < remaining_days then
        get_most_critical (t, None)
      else
        get_most_critical (t, Some cond)
  | ((days, _) as cond) :: t, (Some (curr_days, _) as curr) ->
      if curr_days > days then
        get_most_critical (t, Some cond)
      else
        get_most_critical (t, curr)
in
```

Reply:
> So far this seems a reasonable abstraction. But I'm not sure if this is good for other alerts.

I think now only certificate alert and CSS alert fit to this requirement: message sent to XAPI need to be updated: sent only if not exist, and remove any outdated ones. And there indeed might be other "resource alert", for example, alert for rest amount of disk space, requires the auto-updated requirement.
For your comment, I did tried to add support for the "resource related alert" together with the "expiry alert", but seems it can not be added with little effort: the main reason is that: for the "expiry alert", an "expiry" datetime is needed to generate the "date" field of message_body, while for "resource alert", there will not be the "date" field in the message_body. So I would suggest to only implement for "expiry alert" here, and extend it to support "resource alert" when required later.